rskj-core/src/main/java/co/rsk/rpc/modules/trace/TraceModuleImpl.java

```diff
@@ -204,67 +200,77 @@ private Block getByJsonBlockId(String id) {
             return this.blockchain.getBlockByNumber(0);
         } else if (LATEST_BLOCK.equalsIgnoreCase(id)) {
             return this.blockchain.getBestBlock();
-        } else if ("pending".equalsIgnoreCase(id)) {
+        } else if (PENDING_BLOCK.equalsIgnoreCase(id)) {
             throw RskJsonRpcRequestException.unimplemented("The method don't support 'pending' as a parameter yet");
         } else {
             try {
                 long blockNumber = HexUtils.stringHexToBigInteger(id).longValue();
                 return this.blockchain.getBlockByNumber(blockNumber);
             } catch (NumberFormatException | StringIndexOutOfBoundsException e) {
+                logger.warn("Exception in getBlockByNumber", e);
                 throw RskJsonRpcRequestException.invalidParamError("invalid blocknumber " + id);
             }
         }
     }
 
-    private List<TransactionTrace> buildBlockTraces(Block block) {
+    private Optional<List<TransactionTrace>> buildBlockTraces(Block block) {
         return buildBlockTraces(block, null);
     }
 
-    private List<TransactionTrace> buildBlockTraces(Block block, TraceFilterRequest traceFilterRequest) {
-        List<TransactionTrace> blockTraces = new ArrayList<>();
+    private List<co.rsk.rpc.modules.trace.TransactionTrace> prepareTxTraces(Transaction tx, ProgramTraceProcessor programTraceProcessor, long blockNumber) {
+        TransactionInfo txInfo = receiptStore.getInMainChain(tx.getHash().getBytes(), this.blockStore).orElse(null);
+        Objects.requireNonNull(txInfo);
+        txInfo.setTransaction(tx);
+
+        SummarizedProgramTrace programTrace = (SummarizedProgramTrace) programTraceProcessor.getProgramTrace(tx.getHash());
 
-        if (block != null && block.getNumber() != 0) {
-            List<Transaction> txList = block.getTransactionsList();
+        if (programTrace == null) {
+            return Collections.emptyList();
+        }
 
-            ProgramTraceProcessor programTraceProcessor = new ProgramTraceProcessor();
-            Block parent = this.blockchain.getBlockByHash(block.getParentHash().getBytes());
-            this.blockExecutor.traceBlock(programTraceProcessor, VmConfig.LIGHT_TRACE, block, parent.getHeader(), false, false);
+        return TraceTransformer.toTraces(programTrace, txInfo, blockNumber);
+    }
 
-            if (traceFilterRequest != null) {
-                Stream<Transaction> txStream = block.getTransactionsList().stream();
+    private Optional<List<TransactionTrace>> buildBlockTraces(Block block, TraceFilterRequest traceFilterRequest) {
```

Comment:
@nagarev do you think we could just return `List<TransactionTrace>` instead of `Optional<List<TransactionTrace>>` ? so this may simplify the entire change..

Reply:
Yup! Done!

Reply:
Hey @Vovchyk, had to rethink that due to `empty` being a significant value (not only an empty list) for the code. So I will implement @rmoreliovlabs suggestions to clean that optional a bit at least. 💪🏼